prototype/mobile_endpoint/form/form_processing.py

```diff
@@ -0,0 +1,156 @@
+from datetime import datetime
+from uuid import uuid4, UUID
+
+from flask import logging
+from werkzeug.exceptions import BadRequest
+
+from mobile_endpoint.form.models import XFormInstance, doc_types, XFormDuplicate
+from mobile_endpoint.utils import adjust_datetimes, get_with_lock, ReleaseOnError, LockManager
+import xml2json
+
+
+logger = logging.getLogger(__name__)
+
+INVALID_ID = lambda id: BadRequest('Form submitted with an invalid ID: {}'.format(id))
+
+FORM_PARAMETER_NAME = 'xml_submission_file'
+
+MULTIPART_FILENAME_ERROR = BadRequest((
+    'If you use multipart/form-data, please name your file %s.\n'
+    'You may also do a normal (non-multipart) post '
+    'with the xml submission as the request body instead.\n'
+) % FORM_PARAMETER_NAME)
+MULTIPART_EMPTY_PAYLOAD_ERROR = BadRequest((
+    'If you use multipart/form-data, the file %s'
+    'must not have an empty payload\n'
+) % FORM_PARAMETER_NAME)
+EMPTY_PAYLOAD_ERROR = BadRequest('Post may not have an empty body\n')
+
+
+def get_request_metadata(request):
+    """
+    received_on
+    date_header
+    path
+    submit_ip
+    openrosa_headers
+    """
+    return {
+        'last_sync_token': request.headers.get('last_sync_token', None)
+    }
+
+
+def get_instance_and_attachment(request):
+    attachments = {}
+    if 'multipart/form-data' in request.headers.get('Content-Type'):
+        # ODK submission; of the form
+        # $ curl --form 'xml_submission_file=@form.xml' $URL
+        try:
+            instance = request.files[FORM_PARAMETER_NAME].read().strip()
+        except KeyError:
+            raise MULTIPART_FILENAME_ERROR
+        else:
+            for key, item in request.files.items():
+                if key != FORM_PARAMETER_NAME:
+                    attachments[key] = item
+        if not instance:
+            raise MULTIPART_EMPTY_PAYLOAD_ERROR
+    else:
+        # j2me and touchforms; of the form
+        # $ curl -H 'Content-Type: application/xml' --data '@form.xml' $URL
+        instance = request.data
+        if not instance:
+            raise EMPTY_PAYLOAD_ERROR
+
+    return instance, attachments
+
+
+def create_xform(instance_xml, attachments, request_meta, dao):
```

Comment:
This function is similar to what is in HQ but I changed the way duplicates are handled so that the XML doesn't get re-processed and so that there are less DB queries.